src/v/vlog_error.h

```diff
@@ -0,0 +1,123 @@
+
+class structured_error_code {
+public:
+    constexpr explicit structured_error_code(const char* cat, uint16_t i)
+      : category(cat)
+      , number(i) {}
+
+    const char* category;
+    const uint16_t number;
+};
+} // namespace detail
+
+/**
+ * ======================
+ * Error code definitions
+ * ======================
+ *
+ * These are defined centrally in this file for all subsystems, so that we
+ * can easily avoid accidentally re-using codes for different subsystems.
+ */
+
+/**
+ * Soft assertions are cases that should never happen if our code is bug free,
+ * but which can be recovered and do not necessitate panicking the process.
+ */
+namespace soft_assert {
+constexpr const char* prefix = "ESA";
+
+// We violated internal rules about where start_offset may be set relative
+// to existent segments.
+constexpr auto cloud_storage_start_offset_out_of_range
+  = detail::structured_error_code(prefix, 1);
+
+// We failed to cover all the error types that could occur during a call
+// to object storage.
+constexpr auto cloud_storage_unhandled_client_error
+  = detail::structured_error_code(prefix, 2);
+
+} // namespace soft_assert
+
+/**
+ * I/O errors are cases that indicate an external system is misbehaving in a way
+ * that prevents us from doing our job, for example if a disk gives us an EIO
+ * or if a cloud storage API behaves outside its spec.
+ */
+namespace io_error {
+constexpr const char* prefix = "EIO";
+
+// Operating system indicates I/O error on block device, we cannot proceed.
+constexpr auto local_disk_eio = detail::structured_error_code(prefix, 1);
+
+// Failure reading while doing topic recovery
+constexpr auto cloud_storage_recovery_fatal = detail::structured_error_code(
+  prefix, 2);
+
+// Failure to load data from remote storage to serve a read
+constexpr auto cloud_storage_hydration_fatal = detail::structured_error_code(
+  prefix, 3);
+
+// Backend sent a malformed response body, not compliant with S3 spec
+constexpr auto cloud_storage_bad_response = detail::structured_error_code(
+  prefix, 4);
+
+} // namespace io_error
+
+/**
+ * Consistency errors may indicate corruption in storage.  For example if
+ * a cloud storage manifest refers to a segment that does not exist.  This
+ * can be a bug in redpanda, a bug in the storage backend, or external
+ * interference.
+ */
+namespace consistency_error {
+constexpr const char* prefix = "ECR";
+constexpr auto cloud_storage_missing_segment = detail::structured_error_code(
+  prefix, 1);
+} // namespace consistency_error
+
+/**
+ * Configuration errors indicate that while configuration might have been
+ * syntactically correct, something is wrong with it in practice: for example
+ * we are configured to use an object storage bucket that does not exist.
+ */
+namespace configuration_error {
+constexpr const char* prefix = "ECN";
+constexpr auto cloud_storage_missing_bucket = detail::structured_error_code(
+  prefix, 1);
+
+} // namespace configuration_error
```

Comment:
Categories and numbered constants are nice and elegant, however I don't feel like they bring too much value to offset the inconvinience of their usage. With them, to add a new error case:
* add new error to this file
* use it at the spot

Then, when the error is seen in the log, to see what it means:
* copy the 3-char prefix and find the category in this file
* find the matching code in the category
* search through the code with the code identifier


Compare that to the usage of an uglier old-school approach when each error code is just a short random sequence. To add a new error case:
* generate a random id
* use it at the spot

Then to check a logged error:
* copy the identifier and search through the code with it

The usage could look like
```
vlog_error(klog, "ECN7e01", "....");
```
or even without categories
```
vlog_error(klog, "e79c8293", "....");
```